scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java

```diff
@@ -610,253 +653,356 @@ private void processAlertRecord(InetSocketAddress peerAddress, Record record) th
-						peerConnection = new Connection(record.getPeerAddress());
-
-						Handshaker handshaker = new ResumingServerHandshaker(clientHello.getMessageSeq(),
-								resumableSession, peerConnection, config, maximumTransmissionUnit);
-
-						if (!previousConnection.getPeerAddress().equals(peerConnection.getPeerAddress())) {
-							handshaker.addSessionListener(new SessionAdapter() {
-								@Override
-								public void sessionEstablished(Handshaker handshaker, DTLSSession establishedSession)
-										throws HandshakeException {
-									// when the session will be established for this new connection, remove the previous one.
-									LOGGER.log(Level.FINER,
-											"Remove old connection to [{0}], session with ID [{1}] was successfully resumed on peer [{2}] ",
-											new Object[]{previousConnection.getPeerAddress(),ByteArrayUtils.toHexString(establishedSession.getSessionIdentifier().getId()),establishedSession.getPeer()});
-									terminateConnection(previousConnection, null, null);
-								}
-							});
-						} else {
-							// immediately remove previous connection
-							terminateConnection(previousConnection, null, null);
-						}
-						// add the new one to the store
-						connectionStore.put(peerConnection);
-					
-						// process message
-						nextFlight = handshaker.processMessage(record);
-					} else {
-						LOGGER.log(Level.FINER, "Client [{0}] tries to resume non-existing session with ID [{1}], starting new handshake...",
-								new Object[]{record.getPeerAddress(), ByteArrayUtils.toHexString(sessionId.getId())});
-						// At this point the client has demonstrated reachability by completing a cookie exchange
-						// so we can terminate the previous session.
-						// (see section 4.2.8 of RFC 6347 (DTLS 1.2))
-						if (peerConnection != null) {
-							terminateConnection(peerConnection, null, null);
+				// At this point the client has demonstrated reachability by completing a cookie exchange
+				// so we terminate the previous connection and start a new handshake
+				// (see section 4.2.8 of RFC 6347 (DTLS 1.2))
+				terminateConnection(establishedConnectionProvider.getConnection());
+				startNewHandshake(clientHello, record);
+			}
+		} catch (CookieValidationException e) {
+			sendHelloVerify(clientHello, record, e.getExpectedCookie());
+		}
+	}
+
+	/**
+	 * @param clientHello
+	 * @param record
+	 * @throws CookieValidationException if the sender is not in control of the source IP address
+	 */
+	private void verifyClientControlOfSourceIpAddress(ClientHello clientHello, Record record) {
+		// verify client's ability to respond on given IP address
+		// by exchanging a cookie as described in section 4.2.1 of the DTLS 1.2 spec
+		// see http://tools.ietf.org/html/rfc6347#section-4.2.1
+		byte[] expectedCookie = generateCookie(clientHello);
+		if (!Arrays.equals(expectedCookie, clientHello.getCookie())) {
```

Comment:
return true/false?
This is a potential DoS issue: http://java-performance.info/throwing-an-exception-in-java-is-very-slow/
Exception are slow in general, not a good idea to use it for something else than unfrequent code path